warehouse/utils/security_policy.py

```diff
@@ -0,0 +1,120 @@
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import enum
+
+from pyramid.authorization import Authenticated
+from pyramid.interfaces import ISecurityPolicy
+from pyramid.security import Denied
+from zope.interface import implementer
+
+from warehouse.accounts.interfaces import IUserService
+from warehouse.accounts.models import User
+from warehouse.packaging.models import Project
+
+
+class AuthenticationMethod(enum.Enum):
+    BASIC_AUTH = "basic-auth"
+    SESSION = "session"
+    MACAROON = "macaroon"
+
+
+def principals_for_authenticated_user(userid, request):
+    """Apply the necessary principals to the authenticated user"""
+    login_service = request.find_service(IUserService, context=None)
+    user = login_service.get_user(userid)
+
+    if user is None:
+        return
+
+    if request.session.password_outdated(login_service.get_password_timestamp(userid)):
+        request.session.invalidate()
+        request.session.flash("Session invalidated by password change", queue="error")
+        return
```

Comment:
I believe you're right, but since that's not a bug introduced by this PR I've made a new issue to capture it: https://github.com/pypa/warehouse/issues/11252